packages/contracts/contracts/plugin/PluginSetupProcessor.sol

```diff
@@ -186,394 +246,388 @@ contract PluginSetupProcessor is DaoAuthorizable {
-    function prepareInstallation(
-        address _dao,
-        address _pluginSetup,
-        PluginRepo _pluginSetupRepo,
-        bytes memory _data
-    )
+    /// @return preparedDependency TOD:GIORGI
+    function prepareInstallation(address _dao, PrepareInstallationParams calldata _params)
         external
-        returns (
-            address plugin,
-            address[] memory helpers,
-            PermissionLib.ItemMultiTarget[] memory permissions
-        )
+        returns (address plugin, IPluginSetup.PreparedDependency memory preparedDependency)
     {
+        PluginRepo pluginSetupRepo = _params.pluginSetupRef.pluginSetupRepo;
+
         // Check that the plugin repository exists on the plugin repo registry.
-        if (!repoRegistry.entries(address(_pluginSetupRepo))) {
+        if (!repoRegistry.entries(address(pluginSetupRepo))) {
             revert PluginRepoNonexistent();
         }
 
-        // Reverts if pluginSetup doesn't exist on the repo...
-        _pluginSetupRepo.getVersionByPluginSetup(_pluginSetup);
+        // reverts if not found
+        PluginRepo.Version memory version = pluginSetupRepo.getVersion(
+            _params.pluginSetupRef.versionTag
+        );
 
         // Prepare the installation
-        (plugin, helpers, permissions) = PluginSetup(_pluginSetup).prepareInstallation(_dao, _data);
+        (plugin, preparedDependency) = PluginSetup(version.pluginSetup).prepareInstallation(
+            _dao,
+            _params.data
+        );
+
+        bytes32 pluginInstallationId = _getPluginInstallationId(_dao, plugin);
+
+        bytes32 setupId = _getSetupId(
+            _params.pluginSetupRef,
+            hashPermissions(preparedDependency.permissions),
+            hashHelpers(preparedDependency.helpers),
+            bytes(""),
+            PreparationType.Install
+        );
 
-        // Important safety measure to include dao + plugin manager in the encoding.
-        bytes32 setupId = _getSetupId(_dao, _pluginSetup, address(_pluginSetupRepo), plugin);
+        PluginState storage pluginState = states[pluginInstallationId];
 
-        // Check if the installation was prepared already.
-        if (installPermissionHashes[setupId] != bytes32(0)) {
-            revert SetupAlreadyPrepared();
+        // Allow calling `prepareInstallation` only when
+        // plugin was uninstalled or never been installed before.
+        if (pluginState.currentSetupId != bytes32(0)) {
+            revert PluginAlreadyInstalled();
         }
 
-        installPermissionHashes[setupId] = _getPermissionsHash(permissions);
+        // Only allow to prepare if setupId has not been prepared before.
+        // NOTE that if plugin was uninstalled, the same setupId can still
+        // be prepared as blockNumber would end up being higher than setupId's blockNumber.
+        if (pluginState.blockNumber < pluginState.setupIds[setupId]) {
+            revert SetupAlreadyPrepared(setupId);
+        }
```

Comment:
For this, it doesn't make much sense. It can't be public to be used by outside. So the only use is private function, but in that case, not a big deal, if we leave it like this.